engine-types/src/types/address.rs

```diff
@@ -0,0 +1,144 @@
+use crate::{format, String, TryFrom, H160};
+use borsh::maybestd::io;
+use borsh::{BorshDeserialize, BorshSerialize};
+
+/// Base Eth Address type
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
+pub struct Address(H160);
+
+#[allow(non_snake_case, dead_code)]
+pub const fn ADDRESS(addr: H160) -> Address {
+    Address(addr)
+}
+
+impl Address {
+    /// Construct Address from H160
+    pub fn new(val: H160) -> Self {
+        Self(val)
+    }
+
+    /// Get raw H160 data
+    pub fn raw(&self) -> H160 {
+        self.0
+    }
+
+    /// Encode address to string
+    pub fn encode(&self) -> String {
+        hex::encode(self.0.as_bytes())
+    }
+
+    pub fn decode(address: String) -> Result<Address, error::AddressError> {
+        let data = hex::decode(address).map_err(|_| error::AddressError::FailedDecodeHex)?;
+        if data.len() != 20 {
+            return Err(error::AddressError::IncorrectLength);
+        }
+        let mut result = [0u8; 20];
+        result.copy_from_slice(&data);
+        Ok(Address::new(H160(result)))
+    }
+
+    pub fn as_bytes(&self) -> &[u8] {
+        self.0.as_bytes()
+    }
+
+    pub fn from_slice(raw_addr: &[u8]) -> Self {
+        Self::new(H160::from_slice(raw_addr))
+    }
+
+    pub const fn zero() -> Self {
+        ADDRESS(H160([0u8; 20]))
+    }
+}
+
+impl TryFrom<&[u8]> for Address {
+    type Error = error::AddressError;
+
+    fn try_from(raw_addr: &[u8]) -> Result<Self, Self::Error> {
+        Self::try_from_slice(raw_addr).map_err(|_| error::AddressError::IncorrectLength)
+    }
+}
+
+impl BorshSerialize for Address {
+    fn serialize<W: io::Write>(&self, writer: &mut W) -> io::Result<()> {
+        writer.write_all(self.0.as_bytes())
+    }
+}
+
+impl BorshDeserialize for Address {
+    fn deserialize(buf: &mut &[u8]) -> io::Result<Self> {
+        if buf.len() < 20 {
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                format!("{}", error::AddressError::IncorrectLength),
+            ));
+        }
+        // Guaranty no panics. The length checked early
+        let address = Self(H160::from_slice(&buf[..20]));
+        *buf = &buf[20..];
+        Ok(address)
+    }
+}
+
+impl Default for Address {
+    fn default() -> Self {
+        Address::zero()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_address_serializer() {
```

Comment:
We should also add a test that checks the serialization of `Address` is the same as the serialization of `[u8; 20]`. This will ensure this newtype is backwards compatible with the raw types we were using previously.